### ISSUE TITLE
Add dictionary option

### ISSUE THREAD

**Issue by ashed** [LonamiWebs/Stringlate#62]
Maybe add optional pluggable dictionary, like TranslationMemory in Poedit?

**Comment by XyLoNaMiyX** [comment]
How would you present the suggestions to the users? What suggestions should it load and when? But yeah, I thought this too, sometimes a lot of different projects have a lot of similar strings, or even equal. How would you compare which strings are similar enough to be shown as suggestions?

**Comment by ashed** [comment]
Maybe suggest to user strings with keywords like "sharp", "smooth", "Next", "Previous", etc? And when string contains keyword propose it like suggestion in translation. Keyboard dictionary like in Google Keyboard can help too.

**Comment by XyLoNaMiyX** [comment]
> Maybe suggest to user strings with keywords like "sharp", "smooth", "Next", "Previous", etc?

Could you explain this better? Do you mean like a "downloadable" dictionary for single words, not sentences?

**Comment by ashed** [comment]
Not for the single words, but partially matched sentences.
This is sample of inksape translation ru.po (v0.92)
#: ../inkscape.desktop.in.h:6
msgid "New Drawing" <<< keyword "New"
msgstr "Новый рисунок"
Maybe dictionary must contain original <-->translation couples?

**Comment by XyLoNaMiyX** [comment]
I'm trying to think of a way to extract the "keywords" from the sentences though… First word? Last word? Random word? — Android `string`'s don't have any "keyword" by default (they have a `"name"`, which can or cannot be a valid keyword; and usually are several words_separated_by_underscore). And then with this keyword, Stringlate should look on *all* the translations on other projects to see if it can find a matching result you say?

Sorry for not quite understanding you, I think I'm a bit tired… I just can't think of how this algorithm to detect "keywords" should work, and how should the suggestions be stored. On a database maybe? For each locale though? Or only for the one the user has defined on their phone? Maybe the suggestions could be loaded in a background thread instead… So many questions!

**Comment by ashed** [comment]
Yes, many questions. Sorry for my annoyance.
Google keyboard use user dictionary, which contain wordlist (one word per line, alphabetically normalized).
Maybe background process must scan prepared database, like in GoogleKeyboard? One dict per locale, or one TranslationMemory per project.
Thank you for attention.

**Comment by XyLoNaMiyX** [comment]
> Google keyboard use user dictionary, which contain wordlist

But the problem is that, even with access to this dictionary, there is no such thing like "translation", they're just words which can be suggested, so Stringlate wouldn't know which it should "show". To put an example, the user might have both Spanish and English set on their phones, and the user has self-defined words on both Spanish and English. These have no relation, there is no such thing as "translation". They are just in a database accessible for every application to suggest them.

> Maybe background process must scan prepared database

It *would* be possible to have a database (to provide faster lookup than scanning through the files) containing *all* the strings for *every* locale on *every* project. But again, this is not the problem (only a part of it). The problem is how it should determine the keywords, and how this should be shown to the user on the interface. Maybe as a placeholder? Or as a pop-up when you type? Or an "Apply suggestion"-like button?

Sure, it would be nice to have some kind of natural language processing to determine the best match… But that would be a bit overkill so that's why I was trying to think of a simpler solution.

**Comment by ashed** [comment]
Ok, I'll  connect with Poedit team and talk a little about this problem. Thank you very much.
Besides I translate Stringlate to Russian right now. It is very useful application. Thank you again.

**Comment by XyLoNaMiyX** [comment]
I was thinking of maybe adding the ability to download a dictionary from some website for the currently selected locale and then thought I would need some kind of spell checking… Then realised most keyboards provide this already so it would be fairly useless. What should we do with this issue?

**Comment by XyLoNaMiyX** [comment]
For further reference, it would be good to keep an eye on [Glosbe - the multilingual online dictionary](https://glosbe.com/). If anyone else knows similar projects (besides the ones mentioned on their [/Partners](https://glosbe.com/partners) section), it would be great to know too.

**Comment by Lionirdeadman** [comment]
Dictionaries wouldn't be any good for translations because the user should know their language, what I think could be good is to use Google translate or a similar platform somehow and suggest a translation for the string and the user can decide if the translation string is good or not, this could speed up translation. What do you think @Lonami ?

**Comment by Lonami** [comment]
> Dictionaries wouldn't be any good for translations because the user should know their language,

Sure they *should* know it, but not everyone knows every single word in two languages. A dictionary would be of help for both 1. ensuring you used the right word and 2. not missing those rare words which you don't know.

> use Google translate or a similar platform somehow and suggest a translation for the string and the user can decide if the translation string is good or not

I guess it would be good but what about people if they kept clicking the "Is correct" button without even reading? That's not the idea, we're looking for human quality translations. Using a dictionary would prevent the "auto-click" since sentences are far more complex than a single word.

These are my thoughts, though I think using a translation service which even allowed us to submit back correct translations would be a good help for both too.

**Comment by Lionirdeadman** [comment]
> Sure they should know it, but not everyone knows every single word in two languages. A dictionary would be of help for both 1. ensuring you used the right word and 2. not missing those rare words which you don't know.

Well, I guess it can help but I think it'd be showing too much data to the user if it were to use a dictionary depending on how it's shown. 
> I guess it would be good but what about people if they kept clicking the "Is correct" button without even reading? That's not the idea, we're looking for human quality translations. Using a dictionary would prevent the "auto-click" since sentences are far more complex than a single word.

I think those people could still be using any type of software or none and just write crap anyways. The person checking if those sentences are good should be familiar with the language so they should know if it's actually good or not and change it or not based on that.

**Comment by gsantner** [comment]
Maybe go for full "original string" -> "translation string", and store that in a local db/file. E.g. if theres "settings" -> "Einstellungen" for german, put that in the german translation memory, and if exactly that string comes up, ask if that should be taken

**Comment by Lonami** [comment]
You mean like, loading all the "original string" -> "translated string" from every project in memory, and if an "original string" matches the one from a new project, suggest "translated string"?

**Comment by gsantner** [comment]
Yep

**Comment by gsantner** [comment]
What about using <https://github.com/SQLDroid/SQLDroid/blob/master/README.md> if we go the database direction (which is the better than parsing a file line by line (speed)? It allows us to use normal JDBC to makee sql statements (which is java SE compatible/non-android) while still making use of android internals for database

**Comment by Lonami** [comment]
Although we would have duplicated information (original `.xml` files, needed as a template, and now this database), it would indeed speed up this specific task, so it seems like a good idea to use a database, yes.

**Comment by gsantner** [comment]
I would really like to know how and with which source this guy generates these wordlists <https://github.com/rdoeffinger/Dictionary/releases/tag/v0.2-dictionaries>

**Comment by gsantner** [comment]
But the xml of an project is only loaded when opening a project right?

**Comment by Lonami** [comment]
Yes, opening a project always loads the original `.xml` file(s), and of course, the `.xml` for the selected locale. If you "Peek translations", it will read all the other `.xml` files too (within the same project).
